skimage/filters/rank/core_cy.pyx

```diff
@@ -271,3 +286,293 @@ cdef void _core(void kernel(dtype_t_out*, Py_ssize_t, Py_ssize_t*, double,
+
+    cdef Py_ssize_t centre_p = <Py_ssize_t>(selem.shape[0] / 2) + shift_x
+    cdef Py_ssize_t centre_r = <Py_ssize_t>(selem.shape[1] / 2) + shift_y
+    cdef Py_ssize_t centre_c = <Py_ssize_t>(selem.shape[2] / 2) + shift_z
+
+    # check that structuring element center is inside the element bounding box
+    assert centre_r >= 0
+    assert centre_c >= 0
+    assert centre_p >= 0
+    assert centre_r < srows
+    assert centre_c < scols
+    assert centre_p < splanes
+
+    cdef Py_ssize_t mid_bin = n_bins / 2
+
+    # define pointers to the data
+    cdef char* mask_data = &mask[0, 0, 0]
+
+    # define local variable types
+    cdef Py_ssize_t r, c, p, j, rr, cc, pp, value, local_max, i, even_row
+
+    # number of pixels actually inside the neighborhood (double)
+    cdef double pop = 0
+
+    # build attack and release borders by using difference along axis
+    t = np.dstack((selem, np.zeros((selem.shape[0], selem.shape[1], 1))))
+    cdef unsigned char[:, :, :] t_e = (np.diff(t, axis=2) < 0).view(np.uint8)
+
+    t = np.dstack((np.zeros((selem.shape[0], selem.shape[1], 1)), selem))
+    cdef unsigned char[:, :, :] t_w = (np.diff(t, axis=2) > 0).view(np.uint8)
+
+    t = np.hstack((selem, np.zeros((selem.shape[0], 1, selem.shape[2]))))
+    cdef unsigned char[:, :, :] t_s = (np.diff(t, axis=1) < 0).view(np.uint8)
+
+    t = np.hstack((np.zeros((selem.shape[0], 1, selem.shape[2])), selem))
+    cdef unsigned char[:, :, :] t_n = (np.diff(t, axis=1) > 0).view(np.uint8)
+
+    # the current local histogram distribution
+    cdef Py_ssize_t* histo
+
+    # these lists contain the relative pixel row and column for each of the 4
+    # attack borders east, west, north and south e.g. se_e_r lists the rows of
+    # the east structuring element border
+    cdef Py_ssize_t se_size = srows * scols * splanes * sizeof(Py_ssize_t)
+    cdef Py_ssize_t* se_e_r
+    cdef Py_ssize_t* se_e_c
+    cdef Py_ssize_t* se_e_p
+
+    cdef Py_ssize_t* se_w_r
+    cdef Py_ssize_t* se_w_c
+    cdef Py_ssize_t* se_w_p
+
+    cdef Py_ssize_t* se_n_r
+    cdef Py_ssize_t* se_n_c
+    cdef Py_ssize_t* se_n_p
+
+    cdef Py_ssize_t* se_s_r
+    cdef Py_ssize_t* se_s_c
+    cdef Py_ssize_t* se_s_p
+
+    # number of element in each attack border
+    cdef Py_ssize_t num_se_n, num_se_s, num_se_e, num_se_w, num_se_u, num_se_d
+
+    with nogil:
+
+        se_e_r = <Py_ssize_t*>malloc(se_size)
+        se_e_c = <Py_ssize_t*>malloc(se_size)
+        se_e_p = <Py_ssize_t*>malloc(se_size)
+
+        se_w_r = <Py_ssize_t*>malloc(se_size)
+        se_w_c = <Py_ssize_t*>malloc(se_size)
+        se_w_p = <Py_ssize_t*>malloc(se_size)
+
+        se_n_r = <Py_ssize_t*>malloc(se_size)
+        se_n_c = <Py_ssize_t*>malloc(se_size)
+        se_n_p = <Py_ssize_t*>malloc(se_size)
+
+        se_s_r = <Py_ssize_t*>malloc(se_size)
+        se_s_c = <Py_ssize_t*>malloc(se_size)
+        se_s_p = <Py_ssize_t*>malloc(se_size)
+
+        histo = <Py_ssize_t*>malloc(n_bins * sizeof(Py_ssize_t))
```

Comment:
As above: this would be nicer as a numpy array created outside the loop, which then allows constructs like `histo[:] = 0`, not to mention `histo = np.zeros(n_bins, dtype=np.intp)`. Though, tell me if that causes a big hit on performance. I think most of the performance penalties can be removed with `@cython.boundscheck(False)` and `@cython.wraparound(False)`

Reply:
Same as the above. They are copied from 2D and I can look into them, perhaps changing both versions as well?

Reply:
3D only for now (in this PR).

Reply:
@jni I am having some difficulties making `histo = np.zeros(n_bins, dtype=np.intp)` and `histo[:] = 0` to work - I encountered errors like `Cannot assign type 'long' to 'Py_ssize_t *'`

Reply:
@xialuoke4062 could you push that code so that I can test locally? It can be to a different branch if you want to keep this one clean, or directly to this one — it's easy enough to revert a commit!